src/morphology.rs

```diff
@@ -349,6 +351,631 @@ pub fn close_mut(image: &mut GrayImage, norm: Norm, k: u8) {
+///    99,  99,  99,  99, 222, 222;
+///     7,  99,  99,  99, 222, 222;
+///    99,  99,  99, 222, 222, 222
+/// );
+///
+/// // using a diamond mask
+/// let diamond_eroded = gray_image!(
+///     7,   7,  99,  99,  99,  99;
+///     7,  99,  99,  99,  99,  99;
+///     7,  99,  99,  99,  99, 222;
+///     7,   7,  99,  99,  99, 222;
+///     7,  99,  99,  99, 222, 222
+/// );
+///
+/// assert_pixels_eq!(grayscale_erode(&image, &Mask::diamond(1)), diamond_eroded);
+///
+/// // using a disk mask
+/// let disk_eroded = gray_image!(
+///     7,   7,   7,   7,  99,  99;
+///     7,   7,   7,  99,  99,  99;
+///     7,   7,   7,  99,  99,  99;
+///     7,   7,   7,   7,  99,  99;
+///     7,   7,   7,  99,  99,  99
+/// );
+///
+/// assert_pixels_eq!(grayscale_erode(&image, &Mask::disk(3)), disk_eroded);
+///
+/// // using a square mask
+/// let square_eroded = gray_image!(
+///     7,   7,  99,  99,  99,  99;
+///     7,   7,  99,  99,  99,  99;
+///     7,   7,  99,  99,  99,  99;
+///     7,   7,  99,  99,  99, 222;
+///     7,   7,  99,  99,  99, 222
+/// );
+///
+/// assert_pixels_eq!(grayscale_erode(&image, &Mask::square(1)), square_eroded);
+///
+///
+/// // using an arbitrary mask
+/// // the center of the mask is the '4' here, cause it's coordinates are (0,1)
+/// let column_mask = Mask::from_image(
+///     &gray_image!(
+///         8;
+///         4;
+///         9;
+///         2
+///     ),  
+///     0,  1
+/// );
+///
+/// let column_eroded = gray_image!(
+///     7,  99,  99,  99,  99, 222;
+///     7,  99,  99,  99,  99, 222;
+///     7,  99,  99,  99,  99, 222;
+///     7,  99,  99,  99, 222, 222;
+///     7,  99,  99,  99, 222, 222
+/// );
+///
+/// assert_pixels_eq!(grayscale_erode(&image, &column_mask), column_eroded);
+/// # }
+/// ```
+pub fn grayscale_erode(image: &GrayImage, mask: &Mask) -> GrayImage {
+    #[cfg(feature = "rayon")]
+    let result = GrayImage::from_par_fn(image.width(), image.height(), |x, y| {
+        Luma([mask
+            .apply(image, x, y)
+            .map(|l| l.0[0])
+            .min()
+            .unwrap_or(u8::MAX)]) // default is u8::MIN because it's the neutral element of max
+    });
+    #[cfg(not(feature = "rayon"))]
+    let result = GrayImage::from_fn(image.width(), image.height(), |x, y| {
+        Luma([mask
+            .apply(image, x, y)
+            .map(|l| l.0[0])
+            .min()
+            .unwrap_or(u8::MAX)]) // default is u8::MIN because it's the neutral element of max
+    });
+    result
+}
+
+/// applies a morphologic erosion on the image with the given mask
+///
+/// set each pixel to the minimum of all pixels included in the mask at its position.
+///
+/// See the [`grayscale_erode`](fn.grayscale_erode.html) documentation for examples.
+///
+/// note that unlike `erode_mut`, this performs as many allocations as it's non-mut counterpart
+/// and is thus neither more efficient nor faster
+pub fn grayscale_erode_mut(image: &mut GrayImage, mask: &Mask) {
```

Comment:
If the operations can’t be done in place then there’s no reason for the `_mut` variants - they only exist to let users choose to reuse allocations for functions where this is possible.

Reply:
that makes sense i'll remove them